components/codec/src/byte.rs

```diff
@@ -352,12 +378,172 @@ impl MemComparableCodecHelper for DescendingMemComparableCodecHelper {
     }
 }
 
+pub trait MemComparableByteEncoder: NumberEncoder {
+    /// write all bytes in ascending memory-comparable format.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < values.len().
+    fn write_bytes(&mut self, bs: &[u8]) -> Result<()> {
+        let len = MemComparableByteCodec::encoded_len(bs.len());
+        let buf = unsafe { self.bytes_mut(len) };
+        if buf.len() < len {
+            return Err(Error::BufferTooSmall);
+        }
+        MemComparableByteCodec::encode_all(bs, buf);
+        unsafe {
+            self.advance_mut(len);
+        }
+        Ok(())
+    }
+
+    /// write all bytes in descending memory-comparable format.
+    ///
+    /// # Errors
+    ///
+    /// Returns `Error::BufferTooSmall` if buffer remaining size < values.len().
+    fn write_bytes_desc(&mut self, bs: &[u8]) -> Result<()> {
+        let len = MemComparableByteCodec::encoded_len(bs.len());
+        let buf = unsafe { self.bytes_mut(len) };
+        if buf.len() < len {
+            return Err(Error::BufferTooSmall);
+        }
+        MemComparableByteCodec::encode_all_desc(bs, buf);
+        unsafe {
+            self.advance_mut(len);
+        }
+        Ok(())
+    }
+}
+
+impl<T: NumberEncoder> MemComparableByteEncoder for T {}
+
+pub trait MemComparableByteDecoder: BufferReader {
+    fn read_bytes(&mut self) -> Result<Vec<u8>> {
+        let mut buf = vec![0; self.bytes().len()];
+        let (read, written) = MemComparableByteCodec::try_decode_first(self.bytes(), &mut buf)?;
+        self.advance(read);
+        Ok(buf[..written].to_vec())
+    }
+}
+
+impl<T: BufferReader> MemComparableByteDecoder for T {}
+
+pub trait CompactByteEncoder {
+    /// Joins bytes with its length into a byte slice.
+    /// It is more efficient in both space and time compared to `encode_bytes`.
+    /// Note that the encoded result is not memory-comparable.
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()>;
+}
+
+impl<T: NumberEncoder> CompactByteEncoder for T {
+    #[inline]
+    fn write_compact_bytes(&mut self, data: &[u8]) -> Result<()> {
+        self.write_var_i64(data.len() as i64)?;
+        self.write_all_bytes(data)
+    }
+}
+
+impl CompactByteEncoder for std::fs::File {
```

Comment:
Can we implement it for `std::io::Write`?

Reply:
there will be a conflict if we implement it for `std::io::Write`, since `Vec` is `std::io::Write`,`NumberEnoder`, `CompactByteEncoder`

Reply:
Maybe we can utilize the specialization feature to resolve this by marking the function in BufferWriter as `default fn`: https://github.com/rust-lang/rust/issues/31844 I'm not sure but you can have a try

Reply:
Nevermind. I tried and it won't work. But I still feel uncomfortable here.

@brson Would you like to take a look at this? The background is that we implemented our more efficient codec over buffers, i.e. `T: BufferReader` / `T: BufferWriter` utilizing some buffer-only features, like known allocated size. However, we still need codecs for the non buffer structures, like the traditional `std::fs::File` here.

I think it is too specific to implement `File` for it because `File` is not the only one which cannot utilize buffers. However we cannot directly implement it for `std::io::Write` as well because of duplicate implementations (we have already implemented `CompactByteEncoder` for `T: BufferWriter`). Do you have some other suggestions? Or maybe should we use another trait name, like `CompactByteWriter` for `std::io::Write`?

Reply:
I'll give it a look if it's still an open question, yes.